sdks/python/apache_beam/transforms/util.py

```diff
@@ -815,18 +817,25 @@ def __init__(self, batch_size, max_buffering_duration_secs=None):
     _shard_id_prefix = uuid.uuid4().bytes
 
     def expand(self, pcoll):
+      key_type, value_type = pcoll.element_type.tuple_types
       sharded_pcoll = pcoll | Map(
           lambda key_value: (
               ShardedKey(
                   key_value[0],
                   # Use [uuid, thread id] as the shard id.
                   GroupIntoBatches.WithShardedKey._shard_id_prefix + bytes(
                       threading.get_ident().to_bytes(8, 'big'))),
-              key_value[1]))
+              key_value[1])).with_output_types(
+                  TupleConstraint([ShardedKeyType[key_type], value_type]))
```

Comment:
If you keep this line, I believe the value_type should be `Iterable[value_type]`.

Reply:
Why? The MapFn converts the input K, V to ShardedKey, V. The output of the composite transform is ShardedKey, Iterable[V].

Reply:
Sorry, my mistake I didn't read this carefully enough.